### PR TITLE
Refactor/text

### LINKED ISSUE #1436: text: refactor and add new components
- [x] Heading: adjust size styles, add color prop, remove smallcaps as a size
- [x] Caps: new component with size & color props
- [x] Label: component with size & color props
- [x] P: Add size & color props
- [x] Docs: update docs
- [x] Fix any breaking changes in components

### PR DESCRIPTION
**Pull request by sweeting**
Closes #1436

### CODE REVIEW

File: packages/text/src/vars/index.ts [hunk 1]
```diff
@@ -1,5 +1,19 @@
...
 
-export const headingSizes = keyMirror('smallCaps', 'medium', 'large', 'xLarge')
+export const capsSizes = keyMirror('small', 'medium', 'large')
+
+export const headingSizes = keyMirror(
+  'xsmall',
+  'small',
+  'medium',
+  'large',
+  'xlarge'
```
Comment (jaketrent) [inline packages/text/src/vars/index.ts:10]: Why the change from `xLarge` to `xlarge`?

Offhand, I think that it should remain `xLarge`.  This, because:

- It's a JS key, and camelCase is the preferred convention.
- We treat "extra" as a separate word in camelCasing other var names.

For example, in core: (and more)
```
js/breakpoints.ts
6:  xLarge: '1600px'

js/type.ts
8:  fontSizeXXLarge: '36px',
9:  fontSizeXLarge: '30px',
18:  letterSpacingXXLarge: '-0.72px',
19:  letterSpacingXLarge: '-0.54px',

js/layout.ts
7:  spacingXLarge: '48px',
8:  spacingXXLarge: '64px'
```

File: packages/text/src/vars/index.ts [hunk 2]
```diff
@@ -1,5 +1,19 @@
 import { keyMirror } from '@pluralsight/ps-design-system-util'
 
-export const headingSizes = keyMirror('smallCaps', 'medium', 'large', 'xLarge')
+export const capsSizes = keyMirror('small', 'medium', 'large')
```
Comment (jaketrent) [inline packages/text/src/vars/index.ts:3]: Pretty personal opinion, but the "large" cap size isn't large at all.

For the flexibility of the system in the future, I wonder if we shouldn't choose names that are visually smaller, so that if our design changes in the future where we actually want large caps, we'll have a scale that is easier to adjust in that direction.

Maybe `xXSmall`, `xSmall` and `small`?

File: docs/src/content/components/text-caps.mdx [hunk 3]
```diff
@@ -0,0 +1,92 @@
+---
+name: Caps
+route: /components/text-caps
+---
+
+import { Caps } from '@pluralsight/ps-design-system-text'
+
+# Caps
+
+<Intro>Caps are similar to Label, but all uppercase</Intro>
```
Comment (jaketrent) [inline docs/src/content/components/text-caps.mdx:10]: If the distinguishing feature between Caps and Label is the casing, I wonder why they are not closer aligned in the system.  This could be through naming (eg, `CapsLabel`) or even just properties (eg, `<Label appearance="caps" />`).

File: docs/src/content/components/text-caps.mdx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+```typescript
+import { Caps } from '@pluralsight/ps-design-system-text'
+import React from 'react'
+
+const Example: React.FC = props => {
+  return (
+    <React.Fragment>
+      <div>
+        <Caps color={Caps.colors.secondary}>Secondary</Caps>
```
Comment (jaketrent) [inline docs/src/content/components/text-caps.mdx:61]:
```suggestion
        <Caps color={Caps.colors.primary}>Primar</Caps>
        <Caps color={Caps.colors.secondary}>Secondary</Caps>
```

File: docs/src/content/components/text-caps.mdx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    </React.Fragment>
+  )
+}
+
+export default Example
+```
+
+### Colors
+
+Use the `color` prop to set the color.
```
Comment (jaketrent) [inline docs/src/content/components/text-caps.mdx:51]:
```suggestion
For a more subtle cap label, use `Cap.colors.secondary`.
```

File: docs/src/content/components/text-label.mdx [hunk 6]
```diff
@@ -0,0 +1,97 @@
+---
+name: Label
+route: /components/text-label
+---
+
+import { Label } from '@pluralsight/ps-design-system-text'
+
+# Label
+
+<Intro>Labels are for all text except headings and paragraphs.</Intro>
```
Comment (jaketrent) [inline docs/src/content/components/text-label.mdx:10]: Wow, this is great.  Way to be all-inclusive!

File: docs/src/content/components/text-paragraph.mdx [hunk 7]
```diff
@@ -5,27 +5,93 @@ route: /components/text-paragraph
...
+        consequat.Lorem ipsum dolor sit amet consectetur adipisicing elit, sed
+        do eiusmod tempor incididunt ut labore et dolore magna aliqua.
+      </P>
+      <P size={P.sizes.small}>
+        Small Paragraph - Lorem ipsum dolor sit amet adipisicing elit, sed do
+        eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad
+        minim veniam, quis nostrud exercitation nisi ut aliquip ex ea commodo
+        consequat.Lorem ipsum dolor sit amet consectetur adipisicing elit, sed
+        do eiusmod tempor incididunt ut labore et dolore magna aliqua.
+      </P>
```
Comment (jaketrent) [inline docs/src/content/components/text-paragraph.mdx]:
```suggestion
      <P size={P.sizes.small}>
        *Small Paragraph* - Lorem ipsum dolor sit amet adipisicing elit, sed do
        eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad
        minim veniam, quis nostrud exercitation nisi ut aliquip ex ea commodo
        consequat.Lorem ipsum dolor sit amet consectetur adipisicing elit, sed
        do eiusmod tempor incididunt ut labore et dolore magna aliqua.
      </P>

      <P>
        *Normal Paragraph* - Lorem ipsum dolor sit amet adipisicing elit, sed do
        eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad
        minim veniam, quis nostrud exercitation nisi ut aliquip ex ea commodo
        consequat.Lorem ipsum dolor sit amet consectetur adipisicing elit, sed
        do eiusmod tempor incididunt ut labore et dolore magna aliqua.
      </P>

      <P size={P.sizes.large}>
        *Large Paragraph* - Lorem ipsum dolor sit amet adipisicing elit, sed do
        eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad
        minim veniam, quis nostrud exercitation nisi ut aliquip ex ea commodo
        consequat.Lorem ipsum dolor sit amet consectetur adipisicing elit, sed
        do eiusmod tempor incididunt ut labore et dolore magna aliqua.
      </P>
```

Long example and difficult to visually parse.  Perhaps this will help?

File: docs/src/content/components/text-paragraph.mdx [hunk 8]
```diff
@@ -5,27 +5,93 @@ route: /components/text-paragraph
...
+    </React.Fragment>
+  )
+}
+
+export default Example
+```
+
+### Colors
+
+Use the `color` prop to change the color.
```
Comment (jaketrent) [inline docs/src/content/components/text-paragraph.mdx:57]:
```suggestion
Use the `color` prop to change the color.  Use `P.colors.secondary` for a more subtle paragraph.
```

File: docs/src/content/components/text-paragraph.mdx [hunk 9]
```diff
@@ -5,27 +5,93 @@ route: /components/text-paragraph
...
-      eiusmod tempor incididunt ut labore et dolore magna aliqua.
-    </P>
+    <React.Fragment>
+      <P color={P.colors.secondary}>
+        Paragraph - Lorem ipsum dolor sit amet adipisicing elit, sed do eiusmod
+        tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
+        veniam, quis nostrud exercitation nisi ut aliquip ex ea commodo
+        consequat.Lorem ipsum dolor sit amet consectetur adipisicing elit, sed
+        do eiusmod tempor incididunt ut labore et dolore magna aliqua.
+      </P>
```
Comment (jaketrent) [inline docs/src/content/components/text-paragraph.mdx:72]:
```suggestion
      <P color={P.colors.primary}>
        *Primar color paragraph* - Lorem ipsum dolor sit amet adipisicing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
        veniam, quis nostrud exercitation nisi ut aliquip ex ea commodo
        consequat.Lorem ipsum dolor sit amet consectetur adipisicing elit, sed
        do eiusmod tempor incididunt ut labore et dolore magna aliqua.
      </P>
       
       <P color={P.colors.secondary}>
        *Secondary color paragraph* - Lorem ipsum dolor sit amet adipisicing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
        veniam, quis nostrud exercitation nisi ut aliquip ex ea commodo
        consequat.Lorem ipsum dolor sit amet consectetur adipisicing elit, sed
        do eiusmod tempor incididunt ut labore et dolore magna aliqua.
      </P>
```

File: docs/src/content/components/text-label.mdx [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    </React.Fragment>
+  )
+}
+
+export default Example
+```
+
+### Colors
+
+Use the `color` prop to change the color.
```
Comment (jaketrent) [inline docs/src/content/components/text-label.mdx:54]:
```suggestion
Use the `color` prop to change the color.  Use the `Label.colors.secondary` color for a more subtle label.
```

File: docs/src/content/components/text-label.mdx [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+const Example: React.FC = props => {
+  return (
+    <React.Fragment>
+      <div>
+        <Label color={Caps.colors.secondary}>
+          The technology skills platform
+        </Label>
+      </div>
+    </React.Fragment>
```
Comment (jaketrent) [inline docs/src/content/components/text-label.mdx:68]:
```suggestion
    <>
    <div>
        <Label color={Caps.colors.primary}>
          Primary. The technology skills platform
        </Label>
      </div>
      <div>
        <Label color={Caps.colors.secondary}>
          Secondary. The technology skills platform
        </Label>
      </div>
    </>
```

**Review by jaketrent — CHANGES_REQUESTED** [review]
Brian, this is great work!  This is a large, cohesive dev advancement for this component.  Way to go!

I have made a lot of suggestions inline.  Most are usability suggestions on the docs.

My biggest concern is around the Heading.sizes.xLarge naming change.  That is the issue I'm asking for changes on.

The other thing that I'd invite you to think about is the relationship between Label and Cap.  I have an inline comment wondering if we can make that closer to make their usage (esp for Cap) more clear.

File: packages/text/src/vars/index.ts (hunk 1 above)
Comment (sweeting) [reply packages/text/src/vars/index.ts:10]: @jaketrent I didn't use camelCase in the size prop string values because it seemed they aren't really two words, but rather a name we conjured to represent a size. 

I'm not sure I understand how the string prop value is a JS key and therefore making camelCase it's preferred convention.

If a component were to have a size prop value of 'xxsmall", what would be the correct camelCasing if 'x' is treated as a separate letter: xXSmall?

Comment (jaketrent) [reply packages/text/src/vars/index.ts:10]: @sweeting I see, thank you.  It sounds like you re-cased it to `xlarge` because they didn't seem like two words to you.

As to us having "conjured" the name, I feel like that's too-imaginative a description.  `xLarge` stands for "extra large", which seems easily-inferred enough by abbreviating the "x".

The string value becomes a JS key through the keyMirror utility.  It generates an object that will look like:

```
export const sizes = {
  // ...
  xLarge: 'xLarge'
}
```

See [the implementation of keyMirror](https://github.com/pluralsight/design-system/blob/master/packages/util/src/key-mirror.ts)

In the end, yes, it is a preferred convention, whether it's JS or not.  Also, there is ample precedent for this exact convention for this exact word combo.  In addition to the core examples above, here are more:

```
packages/appframe/src/react/index.tsx
161:    const enteringXLarge = prevLargeMedia !== largeMedia && largeMedia
162:    if (enteringXLarge) openSidenav()

packages/avatar/src/vars/index.ts
5:export const sizes = keyMirror('xSmall', 'small', 'medium', 'large', 'xLarge')
34:  xLarge: '160px'

packages/avatar/src/css/index.ts
33:    fontSize: type.fontSizeXXLarge,
42:  [`.psds-avatar--size-${vars.sizes.xLarge}`]: {
43:    width: vars.widths.xLarge,
44:    height: vars.widths.xLarge,

packages/avatar/src/react/__specs__/__snapshots__/storyshots.spec.ts.snap
156:exports[`Storyshots Default xLarge 1`] = `
161:    alt="xLarge avatar"

packages/datawell/src/css/index.ts
34:    fontSize: type.fontSizeXXLarge,

packages/emptystate/src/css/index.ts
28:    padding: `${layout.spacingXLarge} ${layout.spacingLarge}`,

packages/errors/src/css/index.ts
15:    padding: `${layout.spacingXLarge} ${layout.spacingLarge}`,

packages/layout/src/css/page-width-layout.ts
13:      paddingLeft: layout.spacingXLarge,
14:      paddingRight: layout.spacingXLarge

packages/text/src/css/index.ts
71:  [`.psds-text__heading--size-${vars.headingSizes.xLarge}`]: {
72:    fontSize: type.fontSizeXXLarge,
73:    letterSpacing: type.letterSpacingXLarge,

packages/text/src/vars/index.ts
3:export const headingSizes = keyMirror('smallCaps', 'medium', 'large', 'xLarge')

packages/text/src/react/__stories__/index.story.tsx
16:  <div style={{ padding: core.layout.spacingXLarge }}>{storyFn()}</div>

packages/text/src/react/__specs__/__snapshots__/storyshots.spec.ts.snap
207:exports[`Storyshots Heading size: xLarge 1`] = `
218:    xLarge

packages/tooltip/src/react/__stories__/index.story.tsx
10:  <div style={{ padding: core.layout.spacingXLarge }}>{storyFn()}</div>
```

I am open to changing this to `xlarge` on the merits that exist.  I don't see clear ones yet.  And if we did change it, I'd like to change it consistently across the codebase based on understood rules.  My guess is the ROI on that is not great.

And, to the last question, yes, "extra extra small" would be `xXSmall`.

File: packages/text/src/vars/index.ts (hunk 2 above)
Comment (sweeting) [reply packages/text/src/vars/index.ts:3]: The size names are relative to the category of the type style rather than being relative to all of the type styles. "large" is large for all caps text.

File: docs/src/content/components/text-caps.mdx (hunk 3 above)
Comment (sweeting) [reply docs/src/content/components/text-caps.mdx:10]: Because Caps uses smaller px sizes for the small, medium, large there isn't a 1:1 relationship with Label.

File: packages/text/src/vars/index.ts (hunk 1 above)
Comment (sweeting) [reply packages/text/src/vars/index.ts:10]: I'll change it to align with decision that was made earlier on, but I guess the point I was trying to make is if we abbreviated both 'extra' and 'small' to 'xs' we wouldn't treat both as separate words:

'xXS', 'xS', 's', 'l', 'xL', 'xXL' 

This is probably why numbers are becoming more popular. 🙂

File: packages/text/src/vars/index.ts (hunk 2 above)
Comment (jaketrent) [reply packages/text/src/vars/index.ts:3]: 👍 Totally makes sense.  Was more a forward-looking potential adjustment for if the category was ever expanded.

File: packages/text/src/vars/index.ts (hunk 1 above)
Comment (jaketrent) [reply packages/text/src/vars/index.ts:10]: Great, thank you.

Yes, this is a harder case -- more of an acronym.  Here, there is less consensus in the community at large.  Happens less, practiced less, and thus is easier to not be consistent.  I've seen and written it in all combost: `xxl`, `XXL` and `xXL`.  `HTML` or `DOM` in code are the most common examples that spring to mind.

File: docs/src/content/components/text-caps.mdx (hunk 3 above)
Comment (jaketrent) [reply docs/src/content/components/text-caps.mdx:10]: 👍 Sounds good.

If, however, we called in `CapsLabel` or used `appearance="caps"`, that could remain true.

I was just exploring the relationship between the two: "If they're so close beyond visual style, why are they separate?", kind of thing.

File: docs/src/components/mdx/headings.tsx [hunk 12]
```diff
@@ -9,7 +9,7 @@ export const H1: React.FC<HTMLAttributes<HTMLHeadingElement>> = props => {
   const className = cx(styles.h1, props.className)
 
   return (
-    <Text.Heading size={Text.Heading.sizes.xLarge} className={className}>
+    <Text.Heading size={Text.Heading.sizes.xlarge} className={className}>
```
Comment (EdwardIrby) [inline docs/src/components/mdx/headings.tsx:12]: This change should be reverted looks like xLarge casing didn't change on Heading.Sizes

File: docs/src/components/mdx/headings.tsx [hunk 13]
```diff
@@ -40,7 +40,7 @@ export const H3: React.FC<HTMLAttributes<HTMLHeadingElement>> = props => {
   const className = cx(styles.h3, props.className)
 
   return (
-    <Text.Heading size={Text.Heading.sizes.medium} className={className}>
+    <Text.Heading size={Text.Heading.sizes.xsmall} className={className}>
```
Comment (EdwardIrby) [inline docs/src/components/mdx/headings.tsx:43]: This change should be reverted looks like xSmall casing didn't change on Heading.Sizes

File: docs/src/content/components/text-caps.mdx (hunk 3 above)
Comment (sweeting) [reply docs/src/content/components/text-caps.mdx]: I removed the Caps component and added a boolean prop called caps to the Label component.

File: docs/src/components/mdx/headings.tsx (hunk 12 above)
Comment (sweeting) [reply docs/src/components/mdx/headings.tsx]: Fixed

File: docs/src/components/mdx/headings.tsx (hunk 13 above)
Comment (sweeting) [reply docs/src/components/mdx/headings.tsx]: Fixed

File: docs/src/content/components/text-caps.mdx (hunk 3 above)
Comment (jaketrent) [reply docs/src/content/components/text-caps.mdx]: Oh, cool.  How'd it go?  I'd be interested to hear any of your thinking around why you made the change.

Comment (sweeting) [reply docs/src/content/components/text-caps.mdx]: I ended up with one extra caps size than necessary, but it's probably simpler approach than a separate Caps component. One adjustment I'd like to make is for the `<Labels caps>` font-sizes to be a bit smaller than their non-caps counterparts.

**Review by danethurber — COMMENTED** [review]
@sweeting let's move to the review phase

**Review by jaketrent — COMMENTED** [review]
Another round of feedback that I hope is helpful.

File: docs/src/content/components/text-label.mdx [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+  )
+}
+
+export default Example
+```
+
+### Uppercase label styles
+
+Use all caps text styles sparingly. These can be appropriate for use in badges or in eyebrow headings. Heading styles should be used for headings. Never use these styles for multiple lines of text.
```
[the remark this answers is not in the document]
Comment (sweeting) [reply docs/src/content/components/text-label.mdx:117]: Here are some examples of an eyebrow heading. It's when you combine descriptor text with a more traditional heading style where both can stand on their own.

image

image

image

File: packages/datawell/src/react/index.tsx [hunk 15]
```diff
@@ -55,9 +55,17 @@ const DataWell = React.forwardRef((props, ref) => {
 
   return (
     <div ref={ref} {...styles.dataWell(themeName)} {...rest}>
-      <Heading {...styles.label(themeName)} size={Heading.sizes.smallCaps}>
-        <div>{label}</div>
-      </Heading>
+      <h3>
```
[the remark this answers is not in the document]
Comment (sweeting) [reply packages/datawell/src/react/index.tsx:58]: Replaced with div. I think this was a place where the heading component used to get the uppercase text style threw me off.

File: docs/src/content/components/text-heading.mdx [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+import { Heading } from '@pluralsight/ps-design-system-text'
+import React from 'react'
+
+const Example: React.FC = props => {
+  return (
+    <React.Fragment>
+      <Heading color={Caps.colors.primary}>
+        <h2>Primary heading color</h2>
+      </Heading>
+      <Heading color={Caps.colors.secondary}>
```
[the remark this answers is not in the document]
Comment (sweeting) [reply docs/src/content/components/text-heading.mdx:68]: Fixed

File: docs/src/components/intro.tsx [hunk 17]
```diff
@@ -1,7 +1,8 @@
+import { P } from '@pluralsight/ps-design-system-text'
 import React from 'react'
 
 import * as styles from './intro.module.css'
 
 export const Intro: React.FC = props => {
-  return <div className={styles.intro} {...props} />
+  return <P size={P.sizes.large} {...props} />
```
[the remark this answers is not in the document]
Comment (sweeting) [reply docs/src/components/intro.tsx:7]: Done

File: packages/errors/src/react/error-page.tsx [hunk 18]
```diff
@@ -98,14 +97,16 @@ const ErrorPage: React.FC<ErrorPageProps> = props => {
...
-        <Heading size={headingSizes.medium}>
+        <Heading size={Heading.sizes.medium}>
           <h1>{text}</h1>
         </Heading>
       </div>
       <div {...styles.code(themeName)}>
-        <Heading size={headingSizes.smallCaps}>
-          <h2>Error code: {code}</h2>
-        </Heading>
+        <h2>
```
[the remark this answers is not in the document]
Comment (sweeting) [reply packages/errors/src/react/error-page.tsx:105]: @jaketrent This is a good point. The `<Heading>` component has been indepependent of the `<h1>, <h2>, <h3>, ...` it's used with. Perhaps the `<Heading>` should have numeric levels that correspond to the `<h1>, <h2>, <h3>, ...` tags and then the `<Label>` component could comprise all of the sizes of the typographic scale and be utilized inside the `<Heading>` and `<P>`components.

[the remark this answers is not in the document]
Comment (jaketrent) [reply packages/errors/src/react/error-page.tsx:105]: If `Label` includes all typographic scales/styles, then I think we can do without probably all other Components because a user could just bring their own semantics.  And, in this scenario, maybe it's called something more generic:

```
<Text size="thatExtraBigOneForHeadings">
  <h1>Title</h1>
</Text>

<Text size="thatMidSizeForParas">
  <p>text...</p>
</Text>

<Text size="thatMidSizeForParas">
  <span>label</span>
</Text>

<Text size="thatSmallishForEyebrows" caps>
  <div>Raise your</div>
</Text>

<Text size="thatMidSizeForParas" mono>
  <code>Raise your</code>
</Text>
```

And an alternate syntax which could produce the same thing:

```
<Text.h1 size="thatExtraBigOneForHeadings">
  Title
</Text.h1>

<Text.p size="thatMidSizeForParas">
  text...
</Text.p>

<Text.span size="thatMidSizeForParas">
  label
</Text.span>

<Text.div size="thatSmallishForEyebrows" caps>
  Raise your
</Text.div>

<Text.code size="thatMidSizeForParas">
  Raise your
</Text.code>
```

^ This is a very coder-centric view of the debate. Semantics aren't readily apparent in most product designs.

Regarding the other specific idea of creating 1-6 levels on `Heading` to match `h1-6`.  I believe that this would confuse and combine what we have tried to keep separate historically:

- Heading sizes are in English and only denote visual size
- Markup sizes are 1-6 and only denote semantics

And you will need to mix/match according to the context of your app, which we don't know enough about to anticipate.

File: docs/src/content/components/text-label.mdx (hunk 14 above)
[the remark this answers is not in the document]
Comment (jaketrent) [reply docs/src/content/components/text-label.mdx:117]: awesome, thanks for the edu.

File: docs/src/content/components/text-caps.mdx (hunk 3 above)
Comment (sweeting) [reply docs/src/content/components/text-caps.mdx]: @jaketrent I removed the caps component and added two of the smallest caps styles to heading since that is typically how they are used.

**Review by jaketrent — APPROVED** [review]
Brian, great work!  This has been a long road.  It's one I hope you feel encouraged to walk if you want to.  I think the work is top notch.  Way to go!
